executor/src/tests/executor_tests.rs

```diff
@@ -322,3 +328,120 @@ async fn crash_recovery() {
+            certificate,
+            consensus_index: SequenceNumber::default(),
+        };
+        tx_executor.send(message).await.unwrap();
+    }
+
+    // Feed a 'killer' transaction to the executor. This is a special test transaction that
+    // crashes the test executor engine.
+    let tx0 = 10;
+    let tx1 = KILLER_TRANSACTION;
+    let (digest, batch) = test_batch(vec![tx0, tx1]);
+
+    store.write(digest, batch).await;
+
+    let payload = [(digest, 0)].iter().cloned().collect();
+    let certificate = test_certificate(payload);
+
+    let message = ConsensusOutput {
+        certificate,
+        consensus_index: SequenceNumber::default(),
+    };
+    tx_executor.send(message).await.unwrap();
+
+    // Ensure the execution state is as expected.
+    let _ = rx_output.recv().await;
+    let expected = ExecutionIndices {
+        next_certificate_index: 2,
+        next_batch_index: 0,
+        next_transaction_index: 1,
+    };
+    assert_eq!(execution_state.get_execution_indices().await, expected);
+
+    // Create additional certificates that don't make it to the executor because it crashed.
+    let certificates = test_u64_certificates(
+        /* certificates */ 3, /* batches_per_certificate */ 3,
+        /* transactions_per_batch */ 3,
+    );
+
+    let mut restored_messages = Vec::new();
+    for (certificate, _batches) in certificates {
+        let message = ConsensusOutput {
+            certificate,
+            consensus_index: SequenceNumber::default(),
+        };
+        restored_messages.push(message);
+    }
+    let _ = rx_output.recv().await;
+
+    // Reboot the executor.
+    let (tx_executor, rx_executor) = test_utils::test_channel!(10);
+    let (tx_output, mut rx_output) = channel(10);
+    let (_tx_reconfigure, rx_reconfigure) = watch::channel(reconfigure_notification);
+
+    let _core_handle = Core::spawn(
+        store.clone(),
+        SingleExecutor::new(execution_state.clone(), tx_output),
+        rx_reconfigure,
+        /* rx_subscriber */ rx_executor,
+        restored_messages,
+    );
+
+    // Feed two certificates with good transactions to the executor.
+    let certificates = test_u64_certificates(
+        /* certificates */ 2, /* batches_per_certificate */ 2,
+        /* transactions_per_batch */ 2,
+    );
+    for (certificate, batches) in certificates {
+        for (digest, batch) in batches {
+            store.write(digest, batch).await;
+        }
+        let message = ConsensusOutput {
+            certificate,
+            consensus_index: SequenceNumber::default(),
+        };
+        tx_executor.send(message).await.unwrap();
+    }
+
+    // Ensure the execution state is as expected.
+    let _ = rx_output.recv().await;
+    let expected = ExecutionIndices {
+        next_certificate_index: 7,
+        next_batch_index: 0,
+        next_transaction_index: 0,
+    };
+    assert_eq!(execution_state.get_execution_indices().await, expected);
+}
```

Comment:
This test looks great! Could you please also add a test for the `Executor` as a whole component (not only the `Core` one? At the moment we don't test the recovery part that has to do with the `get_restored_consensus_output` and I believe is essential

Reply:
I looked into testing `get_restored_consensus_output` in isolation, which would involve setting up the consensus and certificate stores, and making sure that the ones sequenced after the certificate index are returned. I think this is better tested manually than via unit test, since the amount of test code required to set up the two databases with state is disproportionate to the two operations in the function.